### PR TITLE
TICKETS #6-#7-#7: Created endpoints for multisig transactions and lis…

### PR DESCRIPTION
**Pull request by pacifiquet**
…ten to events

### CODE REVIEW

File: core/tests/test_event_handler.py [hunk 1]
```diff
@@ -1168,6 +1169,226 @@ def test__fault_proposals(self):
...
+        substrate_service.substrate_interface.generate_multisig_account.return_value = Mock(
+            ss58_address="5H2c4wcccpp7S8PP7HQdzJ5P2DLAGJn6gza3EdQpeQJ5d2Nw"
+        )
+        multisig_account = substrate_service.create_multisig_account(signatories=[signer_one, signer_two], threshold=2)
+        models.MultiSignature.objects.create(
+            address=multisig_account, signatories=[signer_one, signer_two], threshold=2
+        )
+        multisig_address = models.MultiSignature.objects.get(
+            address__iexact="5H2c4wcccpp7S8PP7HQdzJ5P2DLAGJn6gza3EdQpeQJ5d2Nw"
+        )
```
Comment (Shirku) [inline core/tests/test_event_handler.py:1185]: wtf?
```suggestion
        multisig_address = models.MultiSignature.objects.create(
            address=multisig_account, signatories=[signer_one, signer_two], threshold=2
        )
```

File: core/tests/test_event_handler.py [hunk 2]
```diff
@@ -1168,6 +1169,226 @@ def test__fault_proposals(self):
...
+                },
+            },
+        )
+
+        if models.Block.objects.filter(number=0).exists():
+            self.create_new_multisig_event_action(block_new_multisig_event, dao, multisig_address)
+        if models.Block.objects.filter(number=1).exists():
+            self.create_multisig_executed_event_action(block_multisig_executed_event, dao, multisig_address)
+        if models.Block.objects.filter(number=2).exists():
+            self.create_multisig_cancelled_event_action(block__multisig_cancelled_event, dao, multisig_address)
```
Comment (Shirku) [inline core/tests/test_event_handler.py:1300]: i don't understand what's happening here. these should be 3 separate tests, or even better 1 using ddt (lots of examples around).
this fails for me if i run mutliple times, some objects are not cleaned up properly.

File: core/tests/test_models.py [hunk 3]
```diff
@@ -11,3 +11,9 @@ def test_block_str(self):
 
     def test_multiSignature_str(self):
         self.assertEqual(str(models.MultiSignature(pk=5)), "5")
+
+    def test_transaction_str(self):
+        self.assertEqual(str(models.MultisigTransaction(status="PENDING")), "PENDING")
```
Comment (Shirku) [inline core/tests/test_models.py:16]: aside from the fact that this shouldn't be the `__str__` as mentioned in the models.py:
if there are enum values, we should use them.
`status=TransactionStatus.PENDING`

File: core/tests/test_substrate.py [hunk 4]
```diff
@@ -1095,3 +1095,22 @@ def test_create_multisig_account(self):
...
+    def test_create_multisig_event(self):
+        keypair_alice = Keypair.create_from_uri("//Alice")
+        keypair_bob = Keypair.create_from_uri("//Bob")
+        self.si.generate_multisig_account.return_value = Mock(
+            ss58_address="5F3QVbS78a4aTYLiRAD8N3czjqVoNyV42L19CXyhqUMCh4Ch"
+        )
+        multisig_account = self.substrate_service.create_multisig_account([keypair_bob, keypair_alice], 2)
+        self.substrate_service.create_multisig_event(
+            keypair=keypair_alice, multisig_account=multisig_account, value=1, wait_for_inclusion=True
+        )
```
Comment (Shirku) [inline core/tests/test_substrate.py:1108]:
```suggestion
        multisig_account = self.substrate_service.create_multisig_account([keypair_bob, keypair_alice], 2)
        
        self.substrate_service.create_multisig_event(
            keypair=keypair_alice, multisig_account=multisig_account, value=1, wait_for_inclusion=True
        )
```

Comment (Shirku) [reply core/tests/test_substrate.py:1108]: nl to separate (aaa)

File: core/tests/test_views.py [hunk 5]
```diff
@@ -883,3 +883,202 @@ def test_create_multisig_wallet_exist_multsig(self):
...
+            "cancelled_by": None,
+        }
+
+        response = self.client.get(reverse("core-multi-signature-transaction-detail", kwargs={"pk": object_created.pk}))
+        modified_response_data = response.data.copy()
+        modified_response_data.pop("created_at")
+        modified_response_data.pop("updated_at")
+
+        self.assertEqual(response.status_code, HTTP_200_OK)
+        self.assertEqual(modified_response_data, expected_response)
```
Comment (Shirku) [inline core/tests/test_views.py:926]: nl after response.
also this is a bit ugly you can use freezegun for this or mock time

Comment (Shirku) [reply core/tests/test_views.py:926]: same for the other tests

File: core/tests/test_models.py (hunk 3 above)
Comment (Shirku) [reply core/tests/test_models.py:16]: this also applies to other tests e.g. views :P

**Review by Shirku — CHANGES_REQUESTED** [review]
also please always link the ticket in the pr description if possible :P

File: core/models.py [hunk 6]
```diff
@@ -128,3 +128,32 @@ class MultiSignature(Account):
...
+class TransactionStatus(ChoiceEnum):
+    PENDING = "pending"
+    APPROVED = "approved"
+    CANCELLED = "cancelled"
+    EXECUTED = "executed"
+
+
+class MultisigTransaction(TimestampableMixin):
+    multisig = models.ForeignKey(MultiSignature, related_name="transactions", on_delete=models.CASCADE)
+    dao = models.ForeignKey(Dao, related_name="daos", on_delete=models.CASCADE)
```
[the remark this answers is not in the document]
Comment (pacifiquet) [reply core/models.py:142]: in the ticket they mention we need to filter with dao or status

File: core/models.py [hunk 7]
```diff
@@ -128,3 +128,32 @@ class MultiSignature(Account):
...
+    status = models.CharField(max_length=16, choices=TransactionStatus.as_choices(), default=TransactionStatus.PENDING)
+    executed_at = models.DateTimeField(null=True, blank=True)
+    approver = ArrayField(models.CharField(max_length=250))
+    last_approver = models.CharField(max_length=250)
+    cancelled_by = models.CharField(max_length=150, null=True)
```
[the remark this answers is not in the document]
Comment (pacifiquet) [reply core/models.py:147]: approvers and last_approver mentioned in the ticket as well

File: core/event_handler.py [hunk 8]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+            None
+
+        multisignature Transaction based on the Block's events
+        """
+        for new_event in block.event_data.get("Multisig", {}).get("NewMultisig", []):
+            transaction_to_create = []
+            transaction_to_update = []
+            transaction_instances = models.MultisigTransactionOperation.objects.filter(
+                call_hash__in=[new_event.get("call_hash")]
+            )
```
Comment (Shirku) [inline core/event_handler.py:452]: rule of thumb: queries are slow. optimize them if possible. fetch stuff all at once.
this does a new query for each iteration of the for loop. move outside.
in general: the number of queries should not depend on the number of objects (unless we have to do pagination, which we don't here)
please adjust this everyone through this func and add coverage for it in tests.


also what is this? `call_hash__in=[new_event.get("call_hash")]`
why search in a sequence which you create from 1 entry?
`call_hash=new_event.get("call_hash")`

File: core/event_handler.py [hunk 9]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+
+            if transaction_instances:
+                for transaction_instance in transaction_instances:
+                    transaction_instance.last_approver = new_event.get("approving")
+                    transaction_instance.approvers = [new_event.get("approving")]
```
Comment (Shirku) [inline core/event_handler.py:457]: if we know things are always gonna be there we can directly access them e.g.: `new_event["approving"]`

File: core/event_handler.py [hunk 10]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+                    transaction_instance.status = models.TransactionStatus.PENDING
```
Comment (Shirku) [inline core/event_handler.py:458]: so first of all pending appears to be the default value so we don't need to set it.
2ndly do we really want to set a transaction back to pending?

File: core/event_handler.py [hunk 11]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+                    transaction_to_update.append(transaction_instance)
+            else:
```
Comment (Shirku) [inline core/event_handler.py:460]: why else?
there might be some new, some old, no?

File: core/event_handler.py [hunk 12]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+                new_transaction = models.MultisigTransactionOperation(
+                    multisig=models.MultiSignature.objects.filter(address__iexact=new_event.get("multisig")).get(),
+                    dao=models.Dao.objects.filter(owner__address=new_event.get("multisig")).get(),
```
Comment (Shirku) [inline core/event_handler.py:463]: also move out of the loop.
why is there a `get()`?

File: core/event_handler.py (hunk 9 above)
Comment (Shirku) [reply core/event_handler.py:457]: please adjust everywhere where we know it's safe

File: core/event_handler.py [hunk 13]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+            block: Block to multisignature transactions
+
+        Returns:
+            None
+
+        multisignature Transaction based on the Block's events
+        """
+        for new_event in block.event_data.get("Multisig", {}).get("NewMultisig", []):
+            transaction_to_create = []
+            transaction_to_update = []
```
[the remark this answers is not in the document]
Comment (Shirku) [reply core/event_handler.py:449]: names should be plural

File: core/event_handler.py [hunk 14]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+            multi_signature = models.MultiSignature.objects.get(address__iexact=approval_event.get("multisig"))
+            for transaction_instance in (
+                transaction_instances := models.MultisigTransactionOperation.objects.filter(
+                    call_hash__in=[approval_event.get("call_hash")]
+                )
+            ):
+                transaction_instance.last_approver = approval_event.get("approving")
+                transaction_instance.approvers.append(approval_event.get("approving"))
+                transaction_instance.status = models.TransactionStatus.PENDING
+                if len(transaction_instance.approvers) == multi_signature.threshold:
```
Comment (Shirku) [inline core/event_handler.py:489]: `>=`, no?

File: core/event_handler.py [hunk 15]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+
+        for approval_event in block.event_data.get("Multisig", {}).get("MultisigApproval", []):
+            multi_signature = models.MultiSignature.objects.get(address__iexact=approval_event.get("multisig"))
+            for transaction_instance in (
+                transaction_instances := models.MultisigTransactionOperation.objects.filter(
+                    call_hash__in=[approval_event.get("call_hash")]
+                )
+            ):
+                transaction_instance.last_approver = approval_event.get("approving")
+                transaction_instance.approvers.append(approval_event.get("approving"))
```
Comment (Shirku) [inline core/event_handler.py:487]: append? isn't this a list?

File: core/event_handler.py [hunk 16]
```diff
@@ -432,6 +433,87 @@ def _fault_proposals(block: models.Block):
...
+                transaction_instance.status = models.TransactionStatus.PENDING
+                if len(transaction_instance.approvers) == multi_signature.threshold:
+                    transaction_instance.status = models.TransactionStatus.APPROVED
+
+            models.MultisigTransactionOperation.objects.bulk_update(
+                transaction_instances, ("last_approver", "approvers", "status")
+            )
```
Comment (Shirku) [inline core/event_handler.py:494]: we could just update all Transactions at once in the end

File: core/substrate.py [hunk 17]
```diff
@@ -555,6 +555,59 @@ def create_multisig_account(self, signatories: List[str] = None, threshold: int
...
+            call_module=call_module, call_function=call_function, call_params=call_params
+        ).call_hash.hex()
+
+    def create_multisig_event(
+        self,
+        keypair: Keypair,
+        multisig_account,
+        value: int,
+        call_module: str,
+        call_function: str,
```
Comment (Shirku) [inline core/substrate.py:579]: just let the func take an `GenericCall` or smth :P

File: core/substrate.py [hunk 18]
```diff
@@ -555,6 +555,59 @@ def create_multisig_account(self, signatories: List[str] = None, threshold: int
...
+
+        Raises:
+            Any exceptions that occur during the process of submitting the extrinsic.
+        """
+        self.submit_extrinsic(
+            extrinsic=self.substrate_interface.create_multisig_extrinsic(
+                call=self.substrate_interface.compose_call(
+                    call_module=call_module,
+                    call_function=call_function,
+                    call_params={"dest": keypair.ss58_address, "value": value},
```
Comment (Shirku) [inline core/substrate.py:603]: what is `dest`? leftover from the balance thing?

File: core/views.py [hunk 19]
```diff
@@ -255,6 +257,48 @@ def create_multisig(self, request, *args, **kwargs):
...
+    )
+    def create_multisig_transaction(self, request, *args, **kwargs):
+        from core.substrate import substrate_service
+
+        if (
+            call_params := request.data.get("call_params")
+            and (call_function := request.data.get("call_function"))
+            and (call_module := request.data.get("call_module"))
+        ):
+            try:
```
Comment (Shirku) [inline core/views.py:281]: as i said before, please keep `try...except` as small as possible

File: core/views.py [hunk 20]
```diff
@@ -255,6 +257,48 @@ def create_multisig(self, request, *args, **kwargs):
...
+                    call_function=call_function,
+                    call_params=call_params,
+                    status=models.TransactionStatus.PENDING,
+                    multisig=models.MultiSignature.objects.filter(
+                        address__iexact=request.data.get("multisig_address", "")
+                    ).get(),
+                    dao=models.Dao.objects.filter(id__iexact=kwargs.get("pk")).get(),
+                    call_hash=substrate_service.create_transaction_call_hash(
+                        call_params=call_params, call_function=call_function, call_module=call_module
+                    ),
```
Comment (Shirku) [inline core/views.py:293]: why do we create the hash? i thought it should come from FE? https://github.com/deep-ink-ventures/genesis-dao-service/issues/13
```
The first endpoint should store:
the original call data, and
the hash (as calculated by the frontend)
```

File: core/views.py [hunk 21]
```diff
@@ -255,6 +257,48 @@ def create_multisig(self, request, *args, **kwargs):
...
+                )
+                serializer = serializers.RetrieveTransactionOperationSerializer(multisig_transaction)
+                return Response(data=serializer.data, status=HTTP_201_CREATED)
+            except ObjectDoesNotExist:
+                return Response({"message": "Multisignature account or Dao not found."}, HTTP_400_BAD_REQUEST)
```
Comment (Shirku) [inline core/views.py:298]: let's do these separately

File: core/views.py [hunk 22]
```diff
@@ -255,6 +257,48 @@ def create_multisig(self, request, *args, **kwargs):
...
+
+        return Response({"message": "call_module, call_function or call_params are missing."}, HTTP_400_BAD_REQUEST)
```
Comment (Shirku) [inline core/views.py:300]: let's keep following the return early paradigm:
```
if not call_params := request.data.get("call_params")....
        return Response({"message": "call_module, call_function or call_params are missing."}, HTTP_400_BAD_REQUEST)
```

File: core/views.py [hunk 23]
```diff
@@ -367,3 +411,13 @@ class MultiSignatureViewSet(ReadOnlyModelViewSet):
...
+
+
+@method_decorator(
+    swagger_auto_schema(operation_description="Retrieves Transaction Original data and Hash."), "retrieve"
+)
+class MultiSignatureTransactionViewSet(ReadOnlyModelViewSet, SearchableMixin):
+    queryset = models.MultisigTransactionOperation.objects.all()
+    serializer_class = serializers.RetrieveTransactionOperationSerializer
+    allowed_filter_fields = ("dao_id", "status")
+    lookup_field = "pk"
```
Comment (Shirku) [inline core/views.py:423]: that is the default

File: core/substrate.py (hunk 18 above)
Comment (pacifiquet) [reply core/substrate.py:603]: The "dest" key holds the destination address
i should pass it as param as welll

**Pull request merged**